projects/plugins/jetpack/extensions/blocks/google-docs-embed/google-docs-embed.php

```diff
@@ -0,0 +1,151 @@
+			break;
+		case 'google-slides':
+			$pattern = '/^http[s]?:\/\/((?:www\.)?docs\.google\.com(?:.*)?(?:presentation)\/[a-z0-9\/\?=_\-\.\,&%$#\@\!\+]*)\/preview/i';
+			break;
+	}
+
+	if ( empty( $attributes['url'] ) ) {
+		return '';
+	}
+
+	if ( $pattern && ! preg_match( $pattern, $url ) ) {
+		return '';
+	}
+
+	// Add loader for Google Document/Spreadsheets/Presentation blocks.
+	$iframe_markup  = '<iframe src="' . esc_url( $url ) . '" allowFullScreen frameborder="0" title="' . esc_html__( 'Google Document Embed', 'jetpack' ) . '" height="450"></iframe>';
+	$loading_markup = '';
+	$amp_markup     = '';
+
+	if (
+		false !== strpos( $url, '/document/d/' ) ||
+		false !== strpos( $url, '/spreadsheets/d/' ) ||
+		false !== strpos( $url, '/presentation/d/' )
+	) {
+		if ( function_exists( 'amp_is_request' ) && amp_is_request() ) {
+
+			$type = false !== strpos( $url, '/document/d/' ) ? __( 'Google Docs', 'jetpack' ) : '';
+			$type = empty( $type ) && false !== strpos( $url, '/spreadsheets/d/' ) ? __( 'Google Sheets', 'jetpack' ) : $type;
+			$type = empty( $type ) && false !== strpos( $url, '/presentation/d/' ) ? __( 'Google Slides', 'jetpack' ) : $type;
+
+			$iframe_markup = '';
+
+			$amp_markup_message = sprintf(
+				/* translators: Placeholder is a google product, eg. Google Docs, Google Sheets, or Google Slides. */
+				__( 'Tap to open embedded document in %s.', 'jetpack' ),
+				esc_html( $type )
+			);
+
+			$amp_markup = sprintf(
+				'<p class="wp-block-jetpack-google-docs-embed__error-msg"><a target="_blank" rel="noopener noreferrer" href="%s">%s</a></p>',
+				esc_url( $url ),
+				$amp_markup_message
+			);
+
+		} else {
+			$loading_markup = '<div class="loader is-active"><span>' . esc_html__( 'Loading...', 'jetpack' ) . '</span></div>';
+		}
+	}
+
+	$block_classes = Blocks::classes( FEATURE_NAME, $attributes, array( $aspect_ratio ) );
+
+	$html =
+		'<figure class="' . esc_attr( $block_classes ) . '">' .
+			'<div class="wp-block-jetpack-google-docs-embed__wrapper">' .
+				$loading_markup .
+				$amp_markup .
+				$iframe_markup .
+			'</div>' .
+		'</figure>';
+	return $html;
+}
+
+/**
+ * Convert GSuite URL to a preview URL.
+ *
+ * @param string $url The URL of the published Doc/Spreadsheet/Presentation.
+ *
+ * @return string
+ */
+function map_gsuite_url( $url ) {
+
+	// Default regex for all the URLs.
+	$gsuite_regex = '/^(http|https):\/\/(docs\.google.com)\/(spreadsheets|document|presentation)\/d\/([A-Za-z0-9_-]+).*?$/i';
+
+	/**
+	 * Check if the URL is valid.
+	 *
+	 * If not, return the original URL as is.
+	 */
+	preg_match( $gsuite_regex, $url, $matches );
+	if (
+		empty( $matches ) ||
+		empty( $matches[1] ) ||
+		empty( $matches[2] ) ||
+		empty( $matches[3] ) ||
+		empty( $matches[4] )
+	) {
+		return $url;
+	}
+
+	return "{$matches[1]}://$matches[2]/$matches[3]/d/$matches[4]/preview";
+}
```

Comment:
Although this is fairly simple, I think it could be nice to add a test for this function, with a variety of URLs (valid and invalid).